pkg/targetgroupbinding/networking_manager.go

```diff
@@ -225,7 +230,93 @@ func (m *defaultNetworkingManager) consolidateIngressPermissionsPerSGByTGB(_ con
+}
+
+func (m *defaultNetworkingManager) groupIngressPermsBySourceAndProtocolPerSG(_ context.Context) (map[string][]networking.IPPermissionInfo, map[string]map[string]map[string][]networking.IPPermissionInfo) {
+	permsFromIPRangeRulesPerSG := make(map[string][]networking.IPPermissionInfo)
+	permsByProtocolAndSourcePerSG := make(map[string]map[string]map[string][]networking.IPPermissionInfo)
+	for _, ingressPermissionsPerSG := range m.ingressPermissionsPerSGByTGB {
+		for sgID, permissions := range ingressPermissionsPerSG {
+			if _, ok := permsByProtocolAndSourcePerSG[sgID]; !ok {
+				permsByProtocolAndSourcePerSG[sgID] = make(map[string]map[string][]networking.IPPermissionInfo)
+			}
+			for _, permission := range permissions {
+				if len(permission.Permission.UserIdGroupPairs) == 0 && (len(permission.Permission.IpRanges) == 1 || len(permission.Permission.Ipv6Ranges) == 1) {
+					if _, ok := permsFromIPRangeRulesPerSG[sgID]; !ok {
+						permsFromIPRangeRulesPerSG[sgID] = []networking.IPPermissionInfo{}
+					}
+					permsFromIPRangeRulesPerSG[sgID] = append(permsFromIPRangeRulesPerSG[sgID], permission)
+				} else {
+					protocol := awssdk.StringValue(permission.Permission.IpProtocol)
+					if _, ok := permsByProtocolAndSourcePerSG[sgID][protocol]; !ok {
+						permsByProtocolAndSourcePerSG[sgID][protocol] = make(map[string][]networking.IPPermissionInfo)
+					}
+					groupID := ""
+					if len(permission.Permission.UserIdGroupPairs) == 1 {
+						groupID = awssdk.StringValue(permission.Permission.UserIdGroupPairs[0].GroupId)
+					}
+					if _, ok := permsByProtocolAndSourcePerSG[sgID][protocol][groupID]; !ok {
+						permsByProtocolAndSourcePerSG[sgID][protocol][groupID] = []networking.IPPermissionInfo{}
+					}
+					permsByProtocolAndSourcePerSG[sgID][protocol][groupID] = append(permsByProtocolAndSourcePerSG[sgID][protocol][groupID], permission)
+				}
+			}
+		}
+	}
+	return permsFromIPRangeRulesPerSG, permsByProtocolAndSourcePerSG
+}
+
+// computeRestrictedIngressPermissionsPerSG will compute restricted ingress permissions group by source and protocol per SG
+func (m *defaultNetworkingManager) computeRestrictedIngressPermissionsPerSG(ctx context.Context) map[string][]networking.IPPermissionInfo {
+	permsFromIPRangeRulesPerSG, permsByProtocolAndSourcePerSG := m.groupIngressPermsBySourceAndProtocolPerSG(ctx)
+
+	restrictedPermByProtocolPerSG := make(map[string][]networking.IPPermissionInfo)
+	for sgID, permsByProtocolAndSource := range permsByProtocolAndSourcePerSG {
+		if _, ok := restrictedPermByProtocolPerSG[sgID]; !ok {
+			restrictedPermByProtocolPerSG[sgID] = []networking.IPPermissionInfo{}
+		}
+		for _, permsBySource := range permsByProtocolAndSource {
+			for _, perms := range permsBySource {
+				minPort, maxPort := defaultTgbMaxPort, defaultTgbMinPort
+				if len(perms) > 0 {
+					restrictedPermByProtocolPerSG[sgID] = append(restrictedPermByProtocolPerSG[sgID], perms[0])
```

Comment:
We can construct a new Permission object based on perms[0], minPort, maxPort and append to the list at the end of this loop.